js/ui/control/geolocate_control.js

```diff
@@ -1,18 +1,57 @@
 'use strict';
 
 const Evented = require('../../util/evented');
-const browser = require('../../util/browser');
 const DOM = require('../../util/dom');
 const window = require('../../util/window');
 const util = require('../../util/util');
 
 const geoOptions = { enableHighAccuracy: false, timeout: 6000 /* 6sec */ };
 const className = 'mapboxgl-ctrl';
 
+let supportsGeolocation;
+
+function checkGeolocationSupport(callback) {
+
+    if (supportsGeolocation !== undefined) {
+        if (supportsGeolocation === true) {
+            callback();
+        }
+        return;
+    }
+
+    // Test for the case where a browser disables Geolocation because of an
+    // insecure origin
+    // https://sites.google.com/a/chromium.org/dev/Home/chromium-security/deprecating-powerful-features-on-insecure-origins
+    //
+    // navigator.permissions has incomplete browser support
+    // http://caniuse.com/#feat=permissions-api
+    if (window.navigator.permissions !== undefined) {
+        window.navigator.permissions
+            .query({ name:'geolocation' })
+            .then((p) => {
+                supportsGeolocation = p.state !== 'denied';
+                if (supportsGeolocation === true) {
+                    callback();
+                }
+            });
+    } else {
+        supportsGeolocation = !!window.navigator.geolocation;
+        if (supportsGeolocation === true) {
+            callback();
+        }
+    }
+}
```

Comment:
I think it would be slightly cleaner to call the callback in any outcome, and move responsibility for checking `supportsGeolocation` value to `_setupUI`:

```js
function checkGeolocationSupport(callback) {
    if (supportsGeolocation !== undefined) {
        callback();
        
    } else if (window.navigator.permissions !== undefined) {
        window.navigator.permissions.query({name: 'geolocation'}).then((p) => {
            supportsGeolocation = p.state !== 'denied';
            callback();
        });
    } else {
        supportsGeolocation = !!window.navigator.geolocation;
        callback();
    }
}
```